### ISSUE TITLE
Replacing socket.io with uwebsockets for better latency/performance?

### ISSUE THREAD

**Issue by MaZZly** [lance-gg/lance#98]
Hi,

First of all, awesome framework! Finally gotten started writing a game I've wanted to for like 3 years.

I was researching today about scalability of server nodes and stumbled upon this interesting reddit thread:
https://www.reddit.com/r/gamedev/comments/6gjgnx/backend_scaling_for_multiplayer_html5_games/

There they mention that socket.io is not so good and mention that uwebsockets is performing better.

Could this maybe be something worth considering for the websocket communications?

**Comment by namel** [comment]
Hi MaZZly

Good luck with your game.  uwebsockets sounds interesting, and thanks for pointing me to that reddit thread.  

Truth is that I want to move towards UDP, which probably means dropping websockets altogether.  This is what non-javascript (i.e. real) games use, and provides a much better experience.  However last time we checked [UDP was still tricky to use](https://stackoverflow.com/questions/29532157/how-to-talk-to-udp-sockets-with-html5) in node-js environments.

It's very high on the wish-list...

**Comment by MaZZly** [comment]
Yeah ofc better latency is always a plus (as long as it also scales)
 * What are your toughts on browser support? https://caniuse.com/#search=webrtc

Have you checked out https://www.npmjs.com/package/wrtc ? Seems to be handling WebRTC with node?
 * Not sure how much it can handle tho, thinking of scalability.

This might be interesting if you are starting to write the client-server handling with WebRTC:
http://blog.brkho.com/2017/03/15/dive-into-client-server-web-games-webrtc/

---

Found this thread that boasts quite nice numbers for uWS: https://github.com/socketio/engine.io/issues/388
 * Then not sure the project on npm seems to have been deprecated for whatever reason. But I found another one that seems stable https://github.com/ClusterWS/uWS

I was mostly interested in the uWS vs socket.io because there seems to be a benefit CPU-load wise.. And swapping it out doesn't seem like it should be too much work? as uWS was written to be easy to plug in.

**Comment by OpherV** [comment]
@MaZZly The problem is that there are no good node-server based implementation for WebRTC, especially ones that implement unreliable data channels properly. WebRTC in itself was designed to work peer-to-peer rather than with server-client star topology

**Comment by MaZZly** [comment]
So while waiting for something good along the lines of WebRTC to come along, maybe uWebSockets could be put in place? 
https://github.com/uNetworking/uWebSockets.js

**Comment by ouya99** [comment]
I tested http://geckos.io/ . It does WebRTC with UDP channels

Could this be used for lance?

**Comment by MaZZly** [comment]
@OpherV @namel do you have any input for the above Geckos suggestion? I feel like this is quite an important part of getting a multiplayer game to run smoothly with low latencies. And now it seems it would be quite easy to implement? (hopefully) :)
